src/node.rs

```diff
@@ -633,6 +641,27 @@ impl<S: std::fmt::Debug + ForkSource> InMemoryNodeInner<S> {
     pub fn stop_impersonating_account(&mut self, address: Address) -> bool {
         self.impersonated_accounts.remove(&address)
     }
+
+    /// Archives the current state for later queries.
+    pub fn archive_state(&mut self) -> Result<(), String> {
+        if self.previous_states.len() == MAX_PREVIOUS_STATES as usize {
```

Comment:
(let's make it a little bit more defensive - in case someone adds multiple entries to previous_states)

while self.previous_states.len() > MAX_PREVIOUS_STATES